csrc/util_class.cpp

```diff
@@ -0,0 +1,83 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// JNI methods needed by the Java Utils class rather than generic utilities needed by our code.
+
+#include "generated-headers.h"
+
+#include "util.h"
+#include "env.h"
+#include "keyutils.h"
+
+using namespace AmazonCorrettoCryptoProvider;
+
+extern "C" {
+
+/*
+ * Class:     com_amazon_corretto_crypto_provider_Utils
+ * Method:    getNativeBufferOffset
+ * Signature: (Ljava/nio/ByteBuffer;Ljava/nio/ByteBuffer;)J
+ */
+
+JNIEXPORT jlong JNICALL Java_com_amazon_corretto_crypto_provider_Utils_getNativeBufferOffset
+  (JNIEnv *env, jclass, jobject bufA, jobject bufB)
+{
+    jlong no_overlap = (jlong)0x80000000;
+
+    void *pA = env->GetDirectBufferAddress(bufA);
+    void *pB = env->GetDirectBufferAddress(bufB);
+
+    if (!pA || !pB) return no_overlap;
+
+    jlong lenA = env->GetDirectBufferCapacity(bufA);
+    jlong lenB = env->GetDirectBufferCapacity(bufB);
+
+    uintptr_t vA = (uintptr_t)pA;
+    uintptr_t vB = (uintptr_t)pB;
+
+    ptrdiff_t diff = vB - vA;
+    if (diff > 0 && diff >= lenA) {
+        // B is located after A's end, so there's no real overlap
+        return no_overlap;
+    }
+
+    if (diff < 0 && -diff >= lenB) {
+        // A is located after B's end, so no real overlap
+        return no_overlap;
+    }
+
+    // diff should be within jint's bounds now, as direct buffers can't be larger
+    // than can be represented by an int
+    assert(diff < (1L << 31) && diff >= -(1L << 31));
```

Comment:
Should we not be throwing an exception here instead of assert or maybe do both?  If the JVM is built with NDEBUG then this assert will not do anything and we will just return diff.  Is that intended?  I'm not familiar with the build options and best practices inside the JNI layer at Amazon.